MockDuck/Sources/MockRequestResponse.swift

```diff
@@ -22,31 +28,105 @@ final class MockRequestResponse: Codable {
         }
     }
 
-    var response: URLResponse {
-        return responseWrapper.response
+    var response: URLResponse? {
+        return responseWrapper?.response
     }
 
     var responseData: Data? {
         get {
-            return responseWrapper.responseData
+            return responseWrapper?.responseData
         }
         set {
-            responseWrapper.responseData = newValue
+            responseWrapper?.responseData = newValue
         }
     }
 
-    private var requestWrapper: MockRequest
-    private var responseWrapper: MockResponse
+    private(set) lazy var normalizedRequest: URLRequest = {
+        return MockDuck.delegate?.normalizedRequest(for: request) ?? request
+    }()
+
+    let requestWrapper: MockRequest
+    var responseWrapper: MockResponse?
 
     // MARK: - Initializers
 
-    init(request: URLRequest, response: URLResponse, responseData: Data? = nil) {
+    init(request: URLRequest) {
+        self.requestWrapper = MockRequest(request: request)
+        self.responseWrapper = nil
+    }
+
+    init(request: URLRequest, mockResponse: MockResponse) {
+        self.requestWrapper = MockRequest(request: request)
+        self.responseWrapper = mockResponse
+    }
+
+    init(request: URLRequest, response: URLResponse, responseData: Data?) {
         self.requestWrapper = MockRequest(request: request)
         self.responseWrapper = MockResponse(response: response, responseData: responseData)
     }
 
-    convenience init(request: URLRequest, mockResponse: MockResponse) {
-        self.init(request: request, response: mockResponse.response, responseData: mockResponse.responseData)
+    // MARK: - Disk Utilities
+
+    func fileName(for type: MockFileTarget) -> String? {
+        guard let baseName = serializedBaseName else { return nil }
+        let hashValue = serializedHashValue
+        var componentSuffix = ""
+        var pathExtension: String?
+
+        switch type {
+        case .request:
+            pathExtension = "json"
+        case .requestBody:
+            componentSuffix = "-request"
+            pathExtension = request.dataSuffix
+        case .responseData:
+            componentSuffix = "-response"
+            pathExtension = response?.dataSuffix
+        }
+
+        // We only construct a fileName if there is a valid path extension. The only way
+        // pathExtension can be nil here is if this is a data blob that we do not support writing as
+        // an associated file. In this scenario, this data is encoded and stored in the JSON itself
+        // instead of as a separate, associated file.
+        var fileName: String?
+        if let pathExtension = pathExtension {
+            fileName = "\(baseName)-\(hashValue)\(componentSuffix).\(pathExtension)"
+        }
+
+        return fileName
+    }
+
+    var serializedHashValue: String {
```

Comment:
It seems that `serializedHashValue` is duplicated here and in `MockRequest.swift`

Reply:
Good catch! I'll fix.